### PR TITLE
Overhaul CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by akhi3030**
This file was updated in a more-or-less append only manner for some time.  This PR overhauls and update various bits of information presented in the file to make it easier to digest them.

At some point, maybe we should also document our feature stabilisation process in this file.  I figured, I'll restrict this PR to mainly an overhaul and add new content in separate PRs.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,123 +1,126 @@
...
-This process is managed by [cargo-workspaces](https://github.com/pksunkara/cargo-workspaces), with a [bit of magic](https://github.com/pksunkara/cargo-workspaces/compare/master...miraclx:grouping-and-exclusion#files_bucket) sprinkled on top.
-
-## Issue Labels
-
-Issue labels are of the following format `<type>-<content>` where `<type>` is a capital letter indicating the type of the label and `<content>` is a hyphened phrase indicating what is label is about.
-For example, in the label `C-bug`, `C` means category and `bug` means that the label is about bugs.
-Common types include `C`, which means category, `A`, which means area, `T`, which means team.
-
-An issue can have multiple labels including which area it touches, which team should be responsible for the issue, and so on.
-Each issue should have at least one label attached to it after it is triaged and the label could be a general one, such as `C-enhancement` or `C-bug`.
```
Comment (akhi3030) [inline CONTRIBUTING.md:120]: These two sections felt a bit out of place in this file so I dropped them.  I am happy to include them if someone feels strongly that they should be included here.

Comment (matklad) [reply CONTRIBUTING.md:120]: :+1: for trimming this out of CONTRIBUTING.md. CONTRIBUTING.md is a landing page, it should be short and to the point, and just link other documents. 

:-1: for removing this info -- this documents non-obvious institutional knowledge, it should be documented somewhere. It's probably best to find a place for it somewhere in the `docs/` folder. There isn't an obvious place to put it, so I suggest creating a `docs/README.md`. That'll get automatically displayed by github when opening a folder, and can serve both as an index and as dumping ground for stuff we are not quite sure where else to put.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,123 +1,126 @@
...
+Windows is not supported.
+* `nearcore` build process includes generating bindings to RocksDB via
+`bindgen`, which requires `libclang`.  See [`bindgen`
+documentation](https://rust-lang.github.io/rust-bindgen/requirements.html#clang)
+for installation instructions.
+* At the moment, `nearcore` only supports compiling for x86_64 architecture (for
+technical reasons, we have to care about CPU architecture and supporting one is
+easier).  To compile to x86_64 on Apple silicon, use `rustup set default-host
+x86_64-apple-darwin` command.  That is, cross-compilation sadly doesn't work at
+the time of writing.
```
Comment (matklad) [inline CONTRIBUTING.md:71]:
```suggestion
```

We've fixed this!

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,123 +1,126 @@
...
+small code change, please feel free to submit a PR directly addressing the issue
+without opening an issue.
 
-If you see an obvious typo, or an obvious bug that can be fixed with a small change, in the code or documentation, feel free to submit the pull request that fixes it without opening an issue.
+If you have an issue assigned to you, then we kindly ask that you provide an
+update at least once a week so we know if you are still actively working on the
+issue.
 
-## Working on current tasks
+# Setting up editors and IDEs
```
Comment (matklad) [inline CONTRIBUTING.md:39]: Extreme HTML pedantry, but `#` generates an `h1`, and there should be only one `h1` in an article. You want to use `##` for section titeles, and resreve `#` for overall titele of the document.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,123 +1,126 @@
...
+a PR against the `master` branch.
+
+When working on a PR, please keep the following in mind:
+
+1. The changes should be thoroughly tested.  Please refer to [this
+document](https://github.com/nearprotocol/nearcore/wiki/Writing-tests-for-nearcore)
+for our testing guidelines and overview of the testing infrastructure.
+2. If you get an issue assigned to you, please post updates at least once a
+week.  You are encouraged to submit work-in-progress draft PRs to get early
+feedback.
```
Comment (matklad) [inline CONTRIBUTING.md:94]: I suggest cutting this. We are not assigning people to issues/PRs consistenly, and we certainly do let PRs to linger for more than a week. 

To have this rule written down, we need someone to actually enforce it.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,123 +1,126 @@
...
+3. A PR can contain any number of commits and when merged, the commits will be
+squashed into a single commit.
+4. The PR name should follow the template: `<type>: <name>`.  
+Where `type` is `fix` for fixesö `feat` for new features; `refactor` for changes
+that primarily reorganize code; `doc` for changes that primarily change
+documentation or comments; `test` for changes that primarily introduce new
+tests; and `chore` for updating grunt tasks. Please note that `type` is case
+sensitive.
```
Comment (matklad) [inline CONTRIBUTING.md:102]: Let's maybe itemize section titles for bettre readability: 

* `fix`
* `test`
* `feat`
* `chore`

**Review by matklad — APPROVED** [review]
This mostly looks good. What this misses I think is an organized index of links to "further reading", but we don't have organized further reading. Today, this lives in the confluence,

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1,123 +1,126 @@
...
-# Pull Requests and Issues
+- For an overview of the NEAR core architecture, see [this
+playlist](https://www.youtube.com/playlist?list=PL9tzQn_TEuFV4qlts0tVgndnytFs4QSYo).
 
-All the contributions to `nearcore` happen via Pull Requests. To create a Pull Request, fork `nearcore`, create a new branch, do the work there, and then send the PR via the Github interface.
+- If you have never contributed to nearcore before and are looking for
+relatively simple issues to get started with, please checkout issues labeled
+with `C-good-first-issue`
+[here](https://github.com/near/nearcore/labels/C-good-first-issue).  If you see
+one that looks interesting and is not claimed, please comment on the issue that
```
Comment (mina86) [inline CONTRIBUTING.md:18]: I would consider removing ‘and is not claimed’.  We may have issues which are claimed but the author haven’t touched it for months and presumably abandoned it.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -1,123 +1,126 @@
...
+Majority of NEAR developers use CLion with Rust plugin as their primary IDE.
 
-Keep in mind the following:
+We also had success with VSCode with rust-analyzer, see the steps for
+installation
+[here](https://commonwealth.im/near/proposal/discussion/338-remote-development-with-vscode-and-rustanalyzer).
 
-1. The changes need to be thoroughly tested. Refer to [this document](https://github.com/nearprotocol/nearcore/wiki/Writing-tests-for-nearcore) for our testing guidelines and overview of the testing infrastructure.
-2. If you get an issue assigned to you, please post updates at least once a week. It is also preferred for you to send a draft PR as early as you have something working, before it is ready.
+Some of us use VIM with [rust.vim](https://github.com/rust-lang/rust.vim) and
```
Comment (mina86) [inline CONTRIBUTING.md:47]: I feel attacked by Emacs not being mentioned. ;)

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -1,123 +1,126 @@
...
+documentation](https://rust-lang.github.io/rust-bindgen/requirements.html#clang)
+for installation instructions.
+* At the moment, `nearcore` only supports compiling for x86_64 architecture (for
+technical reasons, we have to care about CPU architecture and supporting one is
+easier).  To compile to x86_64 on Apple silicon, use `rustup set default-host
+x86_64-apple-darwin` command.  That is, cross-compilation sadly doesn't work at
+the time of writing.
+* You can optionally use the system installations of `librocksdb`, `libsnappy`
+and `lz4` in order to speed up the compilation and reduce the build memory
+requirements by setting the `ROCKSDB_LIB_DIR`, `SNAPPY_LIB_DIR` and
```
Comment (mina86) [inline CONTRIBUTING.md:74]: Should we link to https://wiki.near.org/contribute/contribute-nearcore#use-system-provided-rocksdb

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -1,123 +1,126 @@
...
+1. The changes should be thoroughly tested.  Please refer to [this
+document](https://github.com/nearprotocol/nearcore/wiki/Writing-tests-for-nearcore)
+for our testing guidelines and overview of the testing infrastructure.
+2. If you get an issue assigned to you, please post updates at least once a
+week.  You are encouraged to submit work-in-progress draft PRs to get early
+feedback.
+3. A PR can contain any number of commits and when merged, the commits will be
+squashed into a single commit.
+4. The PR name should follow the template: `<type>: <name>`.  
+Where `type` is `fix` for fixesö `feat` for new features; `refactor` for changes
```
Comment (mina86) [inline CONTRIBUTING.md:98]:
```suggestion
Where `type` is `fix` for fixes; `feat` for new features; `refactor` for changes
```

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -1,123 +1,126 @@
...
 
+```
 <description>
 
 Test plan
 ---------
 <test plan>
 ```
+6. If your PR introduces a new protocol feature, please document it in
+[CHANGELOG.md](CHANGELOG.md) under `unreleased`.
```
Comment (mina86) [inline CONTRIBUTING.md:113]:
```suggestion
6. If your PR introduces a user-observable change (e.g. a new protocol
feature, new configuration option, new Prometheus metric etc.) please
document it in [CHANGELOG.md](CHANGELOG.md) in `[unreleased]` section.
```

File: CONTRIBUTING.md (hunk 7 above)
Comment (matklad) [reply CONTRIBUTING.md:47]: Yeah... I wouldn't be against dropping this section altogether. We basically use "standard rust tooling", it's not helpful to belabor that. It also can be misleading. For vim, for example, I think the current best source of docs is https://sharksforarms.dev/posts/neovim-rust/ (and, for Emach, https://robert.kra.hn/posts/rust-emacs-setup/, although even that doesn't cover eglot)

File: CONTRIBUTING.md (hunk 4 above)
Comment (akhi3030) [reply CONTRIBUTING.md]: Agreed, now removed.

File: CONTRIBUTING.md (hunk 7 above)
Comment (akhi3030) [reply CONTRIBUTING.md]: Agreed, removed the whole section.

**Comment by akhi3030** [comment]
> This mostly looks good. What this misses I think is an organized index of links to "further reading", but we don't have organized further reading. Today, this lives in the confluence,

Indeed... it will be great to finally make confluence public eventually or move the content elsewhere.

**Comment by akhi3030** [comment]
@matklad , @mina86 : thanks for the great feedback.  I think all comments should be addressed now.  PTAL.

File: CONTRIBUTING.md (hunk 1 above)
Comment (akhi3030) [reply CONTRIBUTING.md:120]: Moved to `docs/README.md`

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -1,136 +1,121 @@
...
 
 ```
-<type>: <name>
-
 <description>
 
 Test plan
 ---------
 <test plan>
 ```
```
Comment (matklad) [inline CONTRIBUTING.md:92]: I'd want to bike shed on whether we need this rule or not. We are definitely not following it, so we should either start enforcing it, or to remove it. But this is out of scope for the PR.

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -1,136 +1,121 @@
...
+## After the PR is submitted
 
-You should expect someone to comment on the issue within 48 hours after it is created. If the proposal in the issue is accepted, you should then follow the process for `Working on current tasks` above.
+1. We have a CI process configured to run various tests on each PR.  All tests
+need to pass before your PR can be merged.
+2. When the CI is green and comments from the reviewer(s) have been addressed,
+they should approve the PR enabling you to merge it.
+3. You can use the `S-automerge` label to instruct Github to put the PR in the
+merge queue with other approved PRs so you do not have to manually look for a
+window to merge it.
```
Comment (matklad) [inline CONTRIBUTING.md:105]: Let's rephrase this a bit maybe:

> To merge an approved PR, add `S-automerge` label. The label is added by PR's author if they have commit access, and by a reviewer otherwise. PR author can also apply automerge label immediaetly after filing a PR: this could remove one round-trip after PR is approved. 

Three bits of info: 

* it's on author to `automerge`
* except that if you don't have commit access
* document pre-emptive auto-merge trick

**Review by matklad — APPROVED** [review]

File: CONTRIBUTING.md (hunk 11 above)
Comment (mina86) [reply CONTRIBUTING.md:92]: I was also suspicious of this.

File: CONTRIBUTING.md (hunk 12 above)
Comment (mina86) [reply CONTRIBUTING.md:105]: Commit access doesn’t tie into it though, does it? Anyone can apply the label, don’t they? It’s just that what usually happens is if we have external contributors they don’t know about it so we end up applying the label.  I’d just change the last point to:

```suggestion
3. You can use the `S-automerge` label to put the PR onto the merge queue with
other approved PRs.
```

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -1,136 +1,121 @@
...
-## Proposing new ideas and features
-
-If you want to propose an idea or a feature and work on it, create a new issue in the `nearcore` repository. We presently do not have an issue template.
+## After the PR is submitted
 
-You should expect someone to comment on the issue within 48 hours after it is created. If the proposal in the issue is accepted, you should then follow the process for `Working on current tasks` above.
+1. We have a CI process configured to run various tests on each PR.  All tests
+need to pass before your PR can be merged.
+2. When the CI is green and comments from the reviewer(s) have been addressed,
+they should approve the PR enabling you to merge it.
```
Comment (mina86) [inline CONTRIBUTING.md:102]:
```suggestion
2. When all the comments from the reviewer(s) have been addressed, they should
approve the PR enabling it to be merged.
```

File: CONTRIBUTING.md (hunk 12 above)
Comment (matklad) [reply CONTRIBUTING.md:105]: >Commit access doesn’t tie into it though, does it?

I think it does: you need to have a triage permission to label issues and PR. Not exactly commit access, but I don't think an external contributor can label even their own PRs.

File: CONTRIBUTING.md (hunk 11 above)
Comment (akhi3030) [reply CONTRIBUTING.md:92]: Agreed, will remove.

File: CONTRIBUTING.md (hunk 12 above)
Comment (akhi3030) [reply CONTRIBUTING.md:98]: Oh, I did not realise external folks cannot access labels.  I suppose it makes sense that they cannot do that.  I suppose if we had a system like [bors](https://github.com/bors-ng/bors-ng) that works via comments, then external folks should be able to merge their own PRs.  Not proposing that we switch but next time we are shopping, we should consider this issue.

I'll update the text.

**Comment by akhi3030** [comment]
@matklad , @mina86 : I assume you guys are happy with me merging this now.  Let me know if you want to do another round of reviews.  Otherwise, I will add the automerge label in a few hours.

**Comment by matklad** [comment]
I am happy! For docs especially, I'd prioritise merging optimistically and submitting follow up PRs if there are concerns!

**Comment by mina86** [comment]
LGTM
